notebooks/06a Modelling - classification.ipynb

```diff
@@ -0,0 +1,835 @@
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define target and training features"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1652704471728
+    }
+   },
+   "outputs": [],
+   "source": [
+    "X = features_df.drop(columns=[\"LENGTH_OF_STAY\"])\n",
+    "y = features_df.risk"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Train/test split\n",
+    "\n",
+    "For model evaluation, we will hold back a 25% test set, and use cross-validation on the remaining 75% for all models until the final comparison is made."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1652704594667
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# Split data for train/test\n",
+    "X_train, X_test, y_train, y_test = train_test_split(\n",
+    "    X, y, train_size=0.75, random_state=42\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Explore class imbalance\n",
+    "\n",
+    "This is a multi-class classification model, so we need to understand where any class imbalance lies otherwise we will skew to larger classes."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Show how many of each class are present in the training set:\n",
+    "X_train.risk.value_counts().sort_index()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Weight samples\n",
+    "\n",
+    "Given significant class imbalance, we will weight samples where the smallest class has a weight of 1, and other classes as a (<1) proportion.\n",
+    "\n",
+    "e.g.\n",
+    "\n",
+    "class|count|weight\n",
+    "---|---|---\n",
+    "1|1000|0.1\n",
+    "2|100|1\n",
+    "3|500|0.2"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "weights = X_train.risk.value_counts().min() / X_train.risk.value_counts().sort_index()\n",
+    "\n",
+    "## Drop risk from training/test set\n",
+    "X_train.drop(columns=\"risk\", inplace=True)\n",
+    "X_test.drop(columns=\"risk\", inplace=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Modeling\n",
+    "\n",
+    "Strategy is to try a number of classification models with:\n",
+    "\n",
+    "* GridsearchCV for hyperparameter tuning with cross validation, refitting full training set for best model\n",
+    "* Test all final models against the held-out test set.\n",
+    "* Explore feature importance of best performing model\n",
+    "* Explore fairness (next notebook) of best performing model\n",
+    "\n",
+    "Logistic regression models are excluded due to significant multi-collinearity (see analysis in Regression notebook)."
```

Comment:
I think we should consider LogReg using L1/L2 regularisation to try and alleviate multicollinearity

Reply:
Will add in with elasticnet